From: Jeffrey Feng <galago1992@gmail.com>
Subject: 
  Add Cairo subpixel rendering backend to poppler-qt5.
  Thanks to Yichao Zhou for his subpixel patch of previous poppler-0.43.
---
git diff --stat
 qt5/demos/viewer.cpp        |  6 +++++-
 qt5/src/CMakeLists.txt      | 14 ++++++++++++++
 qt5/src/poppler-document.cc |  3 +++
 qt5/src/poppler-page.cc     | 46 ++++++++++++++++++++++++++++++++++++++++++++++
 qt5/src/poppler-private.cc  |  2 +-
 qt5/src/poppler-qt5.h       |  3 ++-
 qt5/tests/CMakeLists.txt    |  4 ++++
 7 files changed, 75 insertions(+), 3 deletions(-)

diff --git a/qt5/demos/viewer.cpp b/qt5/demos/viewer.cpp
index d9addc79..ff0269a5 100644
--- a/qt5/demos/viewer.cpp
+++ b/qt5/demos/viewer.cpp
@@ -77,13 +77,17 @@ PdfViewer::PdfViewer(QWidget *parent)
     m_settingsRenderBackendGrp->setExclusive(true);
     act = settingsRenderMenu->addAction(tr("Splash"));
     act->setCheckable(true);
-    act->setChecked(true);
     act->setData(qVariantFromValue(0));
     m_settingsRenderBackendGrp->addAction(act);
     act = settingsRenderMenu->addAction(tr("Arthur"));
     act->setCheckable(true);
     act->setData(qVariantFromValue(1));
     m_settingsRenderBackendGrp->addAction(act);
+    act = settingsRenderMenu->addAction(tr("Cairo"));
+    act->setCheckable(true);
+    act->setChecked(true);
+    act->setData(qVariantFromValue(2));
+    m_settingsRenderBackendGrp->addAction(act);
     connect(m_settingsRenderBackendGrp, &QActionGroup::triggered,
             this, &PdfViewer::slotRenderBackend);
 
diff --git a/qt5/src/CMakeLists.txt b/qt5/src/CMakeLists.txt
index e3b60fcb..3659d467 100644
--- a/qt5/src/CMakeLists.txt
+++ b/qt5/src/CMakeLists.txt
@@ -7,6 +7,10 @@ include_directories(
   ${Qt5Core_INCLUDE_DIRS}
   ${CMAKE_CURRENT_BINARY_DIR}
 )
+if (HAVE_CAIRO)
+  include_directories(${CAIRO_INCLUDE_DIRS})
+  add_definitions(${CAIRO_CFLAGS})
+endif (HAVE_CAIRO)
 
 set(CMAKE_C_VISIBILITY_PRESET hidden)
 set(CMAKE_CXX_VISIBILITY_PRESET hidden)
@@ -38,6 +42,13 @@ set(poppler_qt5_SRCS
   ArthurOutputDev.cc
   poppler-version.cpp
 )
+if (HAVE_CAIRO)
+  set(poppler_qt5_SRCS ${poppler_qt5_SRCS}
+    ${CMAKE_SOURCE_DIR}/poppler/CairoOutputDev.cc
+    ${CMAKE_SOURCE_DIR}/poppler/CairoRescaleBox.cc
+    ${CMAKE_SOURCE_DIR}/poppler/CairoFontEngine.cc
+  )
+endif(HAVE_CAIRO)
 add_library(poppler-qt5 ${poppler_qt5_SRCS})
 set_target_properties(poppler-qt5 PROPERTIES VERSION 1.20.0 SOVERSION 1)
 if(MINGW AND BUILD_SHARED_LIBS)
@@ -45,6 +56,9 @@ if(MINGW AND BUILD_SHARED_LIBS)
     set_target_properties(poppler-qt5 PROPERTIES SUFFIX "-${POPPLER_QT5_SOVERSION}${CMAKE_SHARED_LIBRARY_SUFFIX}")
 endif()
 target_link_libraries(poppler-qt5 poppler ${Qt5Core_LIBRARIES} ${Qt5Gui_LIBRARIES} ${Qt5Xml_LIBRARIES} ${FREETYPE_LIBRARIES})
+if (HAVE_CAIRO)
+  target_link_libraries(poppler-qt5 ${CAIRO_LIBRARIES})
+endif (HAVE_CAIRO)
 if(MSVC)
 target_link_libraries(poppler-qt5 poppler ${poppler_LIBS})
 endif()
diff --git a/qt5/src/poppler-document.cc b/qt5/src/poppler-document.cc
index 81db4eaf..635e44f9 100644
--- a/qt5/src/poppler-document.cc
+++ b/qt5/src/poppler-document.cc
@@ -695,6 +695,9 @@ namespace Poppler {
         ret << Document::SplashBackend;
 #endif
         ret << Document::ArthurBackend;
+#if defined(HAVE_CAIRO)
+        ret << Document::CairoBackend;
+#endif
         return ret;
     }
 
diff --git a/qt5/src/poppler-page.cc b/qt5/src/poppler-page.cc
index bf1f8889..a3684b06 100644
--- a/qt5/src/poppler-page.cc
+++ b/qt5/src/poppler-page.cc
@@ -48,6 +48,7 @@
 #include <QtGui/QPainter>
 
 #include <config.h>
+#include <math.h>
 #include <PDFDoc.h>
 #include <Catalog.h>
 #include <Form.h>
@@ -61,6 +62,9 @@
 #include <SplashOutputDev.h>
 #include <splash/SplashBitmap.h>
 #endif
+#if defined(HAVE_CAIRO)
+#include <CairoOutputDev.h>
+#endif
 
 #include "poppler-private.h"
 #include "poppler-page-transition-private.h"
@@ -622,6 +626,46 @@ QImage Page::renderToImage(double xres, double yres, int xPos, int yPos, int w,
       img = tmpimg;
       break;
     }
+    case Poppler::Document::CairoBackend:
+    {
+#if defined(HAVE_CAIRO)
+      CairoOutputDev *output_dev = new CairoOutputDev();
+      bool subpixel_suited = m_page->page->suitSubpixelRendering (output_dev);
+      if (subpixel_suited) {
+        output_dev->setAntialias(CAIRO_ANTIALIAS_SUBPIXEL);
+      }
+      output_dev->startDoc(m_page->parentDoc->doc);
+
+      // If w or h are -1, that indicates the whole page
+      QSize size = this->pageSize();
+      int buf_width = (w == -1)? qRound( size.width() * xres / 72.0 ) : w;
+      int buf_height = (h == -1)? qRound( size.height() * yres / 72.0 ) : h;
+      // Transform the image if rotate value is specified.
+      int rotate = rotation + m_page->page->getRotate();
+      if (rotate == 90 || rotate == 270) {
+        img = QImage(buf_height, buf_width, QImage::Format_ARGB32);
+      } else {
+        img = QImage(buf_width, buf_height, QImage::Format_ARGB32);
+      }
+      img.fill(Qt::white);  // Never transparent
+
+      cairo_surface_t *surface = cairo_image_surface_create_for_data(
+        img.bits(), CAIRO_FORMAT_ARGB32, img.width(), img.height(),
+        img.bytesPerLine());
+      cairo_t *cairo = cairo_create(surface);
+      output_dev->setCairo(cairo);
+      m_page->parentDoc->doc->displayPageSlice(
+        output_dev, m_page->index + 1, xres, yres, rotation, false, true,
+        false, xPos, yPos, w, h);
+
+      // Clean up
+      output_dev->setCairo(NULL);
+      cairo_destroy(cairo);
+      cairo_surface_destroy(surface);
+      delete output_dev;
+#endif
+      break;
+    }
   }
 
   if (shouldAbortRenderCallback && shouldAbortRenderCallback(payload))
@@ -644,6 +688,8 @@ bool Page::renderToPainter(QPainter* painter, double xres, double yres, int x, i
         QImageDumpingArthurOutputDev arthur_output(painter, nullptr);
         return renderToArthur(&arthur_output, painter, m_page, xres, yres, x, y, w, h, rotate, flags);
     }
+    case Poppler::Document::CairoBackend:
+      return false;
   }
   return false;
 }
diff --git a/qt5/src/poppler-private.cc b/qt5/src/poppler-private.cc
index a3fbec36..05c6bc96 100644
--- a/qt5/src/poppler-private.cc
+++ b/qt5/src/poppler-private.cc
@@ -255,7 +255,7 @@ namespace Debug {
     
     void DocumentData::init()
     {
-        m_backend = Document::SplashBackend;
+        m_backend = Document::CairoBackend;
         paperColor = Qt::white;
         m_hints = 0;
         m_optContentModel = nullptr;
diff --git a/qt5/src/poppler-qt5.h b/qt5/src/poppler-qt5.h
index ebbc456a..37adf71a 100644
--- a/qt5/src/poppler-qt5.h
+++ b/qt5/src/poppler-qt5.h
@@ -1144,7 +1144,8 @@ delete it;
 	*/
 	enum RenderBackend {
 	    SplashBackend,   ///< Splash backend
-	    ArthurBackend   ///< Arthur (Qt) backend
+	    ArthurBackend,   ///< Arthur (Qt) backend
+	    CairoBackend     ///< Cairo backend
 	};
 
 	/**
diff --git a/qt5/tests/CMakeLists.txt b/qt5/tests/CMakeLists.txt
index 5abcbc45..df950efd 100644
--- a/qt5/tests/CMakeLists.txt
+++ b/qt5/tests/CMakeLists.txt
@@ -10,6 +10,10 @@ include_directories(
   ${Qt5Xml_INCLUDE_DIRS}
   ${Qt5Widgets_INCLUDE_DIRS}
 )
+if (HAVE_CAIRO)
+  include_directories(${CAIRO_INCLUDE_DIRS})
+  add_definitions(${CAIRO_CFLAGS})
+endif (HAVE_CAIRO)
 
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${Qt5Core_EXECUTABLE_COMPILE_FLAGS}")
 set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${Qt5Gui_EXECUTABLE_COMPILE_FLAGS}")
